File: frontend/src/lib/api/purchases.ts
import apiClient from './client';

export interface PendingSale {
  id: number;
  numero_pedido: string | null;
  data_criacao: string | null;
  empresa_id: number | null;
  marketplace_id: number | null;
  sku_oferta: string | null;
  nome_produto: string | null;
  quantidade: number | null;
  valor_total_sem_impostos: number | null;
  comissao_sem_impostos: number | null;
  valor_transferido_loja: number | null;
}

export interface PurchaseOrderItem {
  id: number;
  purchase_order_id: number;
  order_id: number;
  sales_order_item_id?: number | null;
  sku_marketplace: string | null;
  sku_fornecedor: string | null;
  quantidade: number;
  custo_unitario: number;
  portes_rateados: number | null;
  impostos_rateados: number | null;
}

export interface GlobalPendingItem {
  id: number;
  numero_pedido: string | null;
  data_criacao: string | null;
  empresa_id: number | null;
  sku_oferta: string | null;
  nome_produto: string | null;
  quantidade: number | null;
  supplier_id: number | null;
  supplier_nome: string | null;
  custo_fornecedor: number | null;
  sku_fornecedor: string | null;
  empresa_nome: string | null;
  valor_transferido_loja: number | null;
}

export interface SaleStateItem {
  order_id: number;
  empresa_id: number | null;
  numero_pedido: string | null;
  sales_order_id: number | null;
  sales_order_item_id: number | null;
  sku_oferta: string | null;
  quantidade: number | null;
  valor_transferido_loja: number | null;
  empresa_nome: string | null;
  estado_venda: string;
  purchase_order_id: number | null;
  po_status: string | null;
  supplier_order_id: string | null;
}

export interface PendingPurchaseItem {
  id: number;
  empresa_id: number;
  sales_order_id: number;
  sales_order_item_id: number;
  sku_marketplace: string | null;
  sku_supplier: string | null;
  supplier_id: number | null;
  quantity: number;
  cost_price_base: number;
  expected_profit: number | null;
  data_criacao: string | null;
  supplier_nome: string | null;
  empresa_nome: string | null;
}

export interface PurchaseOrder {
  id: number;
  empresa_id: number;
  supplier_id: number | null;
  status: string;
  tipo_envio: string;
  total_base: number;
  portes_totais: number;
  impostos: number;
  total_final: number;
  data_criacao: string | null;
  data_ordered: string | null;
  data_paid: string | null;
  notas: string | null;
  billing_nif?: string | null;
  billing_address?: string | null;
  billing_name?: string | null;
  supplier_order_id?: string | null;
  tracking_number?: string | null;
  supplier_nome: string | null;
  empresa_nome?: string | null;
  order_refs?: string | null;
  items?: PurchaseOrderItem[];
}

/** PO com estado de fatura e conta corrente (match PO ↔ Fatura) */
export interface PurchaseOrderWithInvoiceStatus {
  id: number;
  empresa_id: number;
  supplier_id: number | null;
  status: string;
  tipo_envio: string;
  total_final: number;
  data_criacao: string | null;
  data_ordered: string | null;
  invoice_ref: string | null;
  invoice_amount: number | null;
  invoice_pdf_url: string | null;
  supplier_nome: string | null;
  empresa_nome: string | null;
  has_ledger_entry: boolean;
  invoice_status: 'none' | 'url_only' | 'ref_and_amount';
  is_reconciled?: boolean;
}

/** v3.5: Detalhe para Checkout Inteligente (custos, IVA por regime, margem) */
export interface CheckoutDetail {
  id: number;
  empresa_id: number;
  supplier_id: number | null;
  status: string;
  tipo_envio: string;
  office_id: number | null;
  total_base: number;
  portes_totais: number;
  impostos: number;
  total_final: number;
  taxas_pagamento: number;
  valor_base_artigos: number;
  iva_total: number;
  custo_portes_fornecedor: number;
  billing_nif: string | null;
  billing_address: string | null;
  billing_name: string | null;
  supplier_order_id: string | null;
  notas: string | null;
  supplier_nome: string | null;
  supplier_designacao: string | null;
  supplier_nif: string | null;
  supplier_morada: string | null;
  supplier_cp: string | null;
  supplier_localidade: string | null;
  supplier_pais: string | null;
  regime_iva: string | null;
  empresa_nome: string | null;
  empresa_nif: string | null;
  empresa_morada: string | null;
  empresa_pais: string | null;
  taxa_iva_pct: number;
  linked_sales_total: number;
  margin_eur: number;
  margin_pct: number;
  shipping_address: { tipo: string; designacao: string; morada?: string; codigo_postal?: string; localidade?: string; pais?: string };
  items: Array<{
    id: number;
    purchase_order_id: number;
    sales_order_item_id: number | null;
    sku_marketplace: string | null;
    sku_fornecedor: string | null;
    quantidade: number;
    custo_unitario: number;
    portes_rateados: number;
    impostos_rateados: number;
  }>;
}

export interface AggregateResponse {
  success: boolean;
  purchase_order_id: number | null;
  total_base: number | null;
  portes_totais: number | null;
  impostos: number | null;
  total_final: number | null;
  items_count: number | null;
  error: string | null;
}

export const purchasesApi = {
  getPendingSales: async (empresaId: number, limit = 500): Promise<{ items: PendingSale[] }> => {
    const { data } = await apiClient.get<{ items: PendingSale[] }>(
      `/api/v1/purchases/pending-sales?empresa_id=${empresaId}&limit=${limit}`
    );
    return data;
  },

  listPurchaseOrders: async (
    empresaId?: number,
    status?: string,
    limit = 100,
    offset = 0,
    dataInicio?: string,
    dataFim?: string,
  ): Promise<{ items: PurchaseOrder[]; total: number }> => {
    const params = new URLSearchParams();
    if (empresaId != null) params.append('empresa_id', String(empresaId));
    if (status) params.append('status', status);
    if (dataInicio) params.append('data_inicio', dataInicio);
    if (dataFim) params.append('data_fim', dataFim);
    params.append('limit', String(limit));
    params.append('offset', String(offset));
    const { data } = await apiClient.get<{ items: PurchaseOrder[]; total: number }>(
      `/api/v1/purchases/orders?${params}`
    );
    return data;
  },

  getPurchaseOrder: async (purchaseOrderId: number): Promise<PurchaseOrder> => {
    const { data } = await apiClient.get<PurchaseOrder>(`/api/v1/purchases/orders/${purchaseOrderId}`);
    return data;
  },

  /** v3.5: Checkout Inteligente — detalhe com custos, IVA por regime, margem */
  getCheckoutDetail: async (
    purchaseOrderId: number,
    params?: { portes?: number; taxas_pagamento?: number }
  ): Promise<CheckoutDetail> => {
    const sp = new URLSearchParams();
    if (params?.portes != null) sp.append('portes', String(params.portes));
    if (params?.taxas_pagamento != null) sp.append('taxas_pagamento', String(params.taxas_pagamento));
    const q = sp.toString();
    const { data } = await apiClient.get<CheckoutDetail>(
      `/api/v1/purchases/orders/${purchaseOrderId}/checkout-detail${q ? `?${q}` : ''}`
    );
    return data;
  },

  updatePoTotals: async (
    purchaseOrderId: number,
    body: { portes_totais: number; taxas_pagamento: number; total_final?: number; valor_base_artigos?: number; iva_total?: number }
  ): Promise<{ success: boolean; total_final?: number; margin_eur?: number; margin_pct?: number }> => {
    const { data } = await apiClient.patch(
      `/api/v1/purchases/orders/${purchaseOrderId}/totals`,
      body
    );
    return data;
  },

  /** Altera o fornecedor da PO (checkout: escolher outro fornecedor que não o proposto). */
  setPoSupplier: async (purchaseOrderId: number, supplierId: number): Promise<{ success: boolean; supplier_id?: number }> => {
    const { data } = await apiClient.patch(
      `/api/v1/purchases/orders/${purchaseOrderId}/supplier`,
      { supplier_id: supplierId }
    );
    return data;
  },

  finalizePo: async (
    purchaseOrderId: number,
    body?: { supplier_order_id?: string; portes_totais?: number; taxas_pagamento?: number; total_final?: number; valor_base_artigos?: number; iva_total?: number }
  ): Promise<{ success: boolean; status?: string }> => {
    const { data } = await apiClient.post(
      `/api/v1/purchases/orders/${purchaseOrderId}/finalize`,
      body ?? {}
    );
    return data;
  },

  updateStatus: async (purchaseOrderId: number, status: string): Promise<{ success: boolean }> => {
    const { data } = await apiClient.patch<{ success: boolean }>(
      `/api/v1/purchases/orders/${purchaseOrderId}`,
      { status }
    );
    return data;
  },

  aggregate: async (body: {
    empresa_id: number;
    order_ids: number[];
    supplier_id?: number;
    tipo_envio?: string;
    portes_totais?: number;
    taxa_iva_pct?: number;
  }): Promise<AggregateResponse> => {
    const { data } = await apiClient.post<AggregateResponse>('/api/v1/purchases/aggregate', body);
    return data;
  },

  getGlobalPending: async (supplierId?: number, limit = 1000): Promise<{ items: GlobalPendingItem[] }> => {
    const params = new URLSearchParams();
    if (supplierId != null) params.append('supplier_id', String(supplierId));
    params.append('limit', String(limit));
    const { data } = await apiClient.get<{ items: GlobalPendingItem[] }>(`/api/v1/purchases/global-pending?${params}`);
    return data;
  },

  prepareBulk: async (body: {
    order_ids: number[];
    portes_totais?: number;
    taxa_iva_pct?: number;
    tipo_envio?: string;
  }): Promise<{ success: boolean; error?: string; purchase_orders: { purchase_order_id: number; empresa_id: number; supplier_id: number | null; items_count: number; total_final: number; billing_name: string | null }[] }> => {
    const { data } = await apiClient.post('/api/v1/purchases/prepare-bulk', body);
    return data;
  },

  setSupplierOrderId: async (purchaseOrderId: number, supplierOrderId: string): Promise<{ success: boolean }> => {
    const { data } = await apiClient.patch<{ success: boolean }>(
      `/api/v1/purchases/orders/${purchaseOrderId}/supplier-order-id`,
      { supplier_order_id: supplierOrderId }
    );
    return data;
  },

  /** Lista POs com estado de fatura e conta corrente (para match PO ↔ Fatura) */
  listPurchaseOrdersWithInvoiceStatus: async (
    empresaId?: number,
    status?: string,
    invoiceState?: string,
    limit = 200,
    offset = 0
  ): Promise<{ items: PurchaseOrderWithInvoiceStatus[]; total: number }> => {
    const params = new URLSearchParams();
    if (empresaId != null) params.append('empresa_id', String(empresaId));
    if (status) params.append('status', status);
    if (invoiceState) params.append('invoice_state', invoiceState);
    params.append('limit', String(limit));
    params.append('offset', String(offset));
    params.append('with_invoice_status', 'true');
    const { data } = await apiClient.get<{ items: PurchaseOrderWithInvoiceStatus[]; total: number }>(
      `/api/v1/purchases/orders?${params}`
    );
    return data;
  },

  /** Regista fatura na PO e opcionalmente envia para conta corrente */
  registerPoInvoice: async (
    purchaseOrderId: number,
    body: { invoice_ref?: string; invoice_amount?: number; post_to_ledger?: boolean }
  ): Promise<{ success: boolean; ledger_created?: boolean }> => {
    const { data } = await apiClient.patch<{ success: boolean; ledger_created?: boolean }>(
      `/api/v1/purchases/orders/${purchaseOrderId}/invoice`,
      { invoice_ref: body.invoice_ref ?? null, invoice_amount: body.invoice_amount ?? null, post_to_ledger: body.post_to_ledger !== false }
    );
    return data;
  },

  getSaleState: async (empresaId?: number, limit = 500): Promise<{ items: SaleStateItem[] }> => {
    const params = new URLSearchParams();
    if (empresaId != null) params.append('empresa_id', String(empresaId));
    params.append('limit', String(limit));
    const { data } = await apiClient.get<{ items: SaleStateItem[] }>(`/api/v1/purchases/sale-state?${params}`);
    return data;
  },

  /** Fase 3: Central de Compras — pending_purchase_items (todas as empresas) */
  getPendingForCockpit: async (supplierId?: number, limit = 2000): Promise<{ items: PendingPurchaseItem[] }> => {
    const params = new URLSearchParams();
    if (supplierId != null) params.append('supplier_id', String(supplierId));
    params.append('limit', String(limit));
    const { data } = await apiClient.get<{ items: PendingPurchaseItem[] }>(`/api/v1/purchases/pending?${params}`);
    return data;
  },

  /** Fase 3: Split fiscal — cria POs por (empresa_id, supplier_id) a partir de pending_item_ids */
  consolidatePurchases: async (body: {
    pending_item_ids: number[];
    portes_totais?: number;
    taxa_iva_pct?: number;
  }): Promise<{
    success: boolean;
    error?: string;
    purchase_orders: { purchase_order_id: number; empresa_id: number; supplier_id: number | null; items_count: number; total_final: number; billing_name: string | null }[];
    num_pos?: number;
    num_items_processed?: number;
  }> => {
    const { data } = await apiClient.post('/api/v1/purchases/consolidate', body);
    return data;
  },

  /** Fase 3: Lista de POs em Draft (para Wizard de Checkout) */
  getDrafts: async (empresaId?: number, limit = 100, offset = 0): Promise<{ items: PurchaseOrder[]; total: number; limit: number; offset: number }> => {
    const params = new URLSearchParams();
    if (empresaId != null) params.append('empresa_id', String(empresaId));
    params.append('limit', String(limit));
    params.append('offset', String(offset));
    const { data } = await apiClient.get<{ items: PurchaseOrder[]; total: number; limit: number; offset: number }>(`/api/v1/purchases/drafts?${params}`);
    return data;
  },
};

File: backend/app/api/v1/purchases.py
"""Endpoints do Módulo de Compras (Purchases): draft automático, consolidação, rentabilidade triangular."""
from fastapi import APIRouter, Query, Depends, HTTPException
from typing import List, Optional
from pydantic import BaseModel
from app.services.purchase_service import (
    SKUTranslatorService,
    OrderAggregatorService,
    TriangularProfitService,
    PurchaseOrderService,
    BulkPurchaseService,
)
from app.services.purchase_aggregator_service import PurchaseAggregatorService
from app.services.purchase_checkout_service import PurchaseCheckoutService
from app.services.accounting_match_service import AccountingMatchService
from app.api.deps import get_bulk_purchase_service, get_purchase_aggregator_service

router = APIRouter(prefix="/purchases", tags=["purchases"])


class SKUResolveResponse(BaseModel):
    sku_mapping_id: Optional[int]
    supplier_id: Optional[int]
    sku_fornecedor: Optional[str]
    custo_fornecedor: float
    nome_produto: Optional[str]
    supplier_nome: Optional[str]


class AggregateRequest(BaseModel):
    empresa_id: int
    order_ids: List[int]
    supplier_id: Optional[int] = None
    tipo_envio: str = "Escritorio"
    portes_totais: float = 0
    taxa_iva_pct: float = 0


class AggregateResponse(BaseModel):
    success: bool
    purchase_order_id: Optional[int] = None
    total_base: Optional[float] = None
    portes_totais: Optional[float] = None
    impostos: Optional[float] = None
    total_final: Optional[float] = None
    items_count: Optional[int] = None
    error: Optional[str] = None


@router.get("/sku-resolve", response_model=SKUResolveResponse)
async def resolve_sku(
    empresa_id: int = Query(...),
    sku_marketplace: str = Query(...),
    marketplace_id: Optional[int] = Query(None),
):
    """Dado SKU do marketplace, devolve fornecedor, SKU fornecedor e custo base (SKU Translator)."""
    svc = SKUTranslatorService()
    try:
        r = svc.resolve(empresa_id, sku_marketplace, marketplace_id)
        if not r:
            return SKUResolveResponse(custo_fornecedor=0)
        return SKUResolveResponse(**r)
    finally:
        svc.close()


@router.post("/aggregate", response_model=AggregateResponse)
async def create_purchase_order_from_sales(body: AggregateRequest):
    """
    Agrega várias vendas (order_ids) numa única purchase_order (Draft).
    Order Aggregator: consolidação 1:N com rateio de portes e IVA.
    """
    svc = OrderAggregatorService()
    try:
        result = svc.create_draft_from_sales(
            empresa_id=body.empresa_id,
            order_ids=body.order_ids,
            supplier_id=body.supplier_id,
            tipo_envio=body.tipo_envio,
            portes_totais=body.portes_totais,
            taxa_iva_pct=body.taxa_iva_pct,
        )
        return AggregateResponse(
            success=result["success"],
            purchase_order_id=result.get("purchase_order_id"),
            total_base=result.get("total_base"),
            portes_totais=result.get("portes_totais"),
            impostos=result.get("impostos"),
            total_final=result.get("total_final"),
            items_count=result.get("items_count"),
            error=result.get("error"),
        )
    finally:
        svc.close()


@router.get("/rentabilidade")
async def get_rentabilidade_triangular(
    empresa_id: Optional[int] = Query(None),
    purchase_order_id: Optional[int] = Query(None),
    order_id: Optional[int] = Query(None),
):
    """Vista de rentabilidade triangular: Venda vs Compra. Margem real por linha (com rateio se consolidado)."""
    svc = TriangularProfitService()
    try:
        rows = svc.get_rentabilidade(empresa_id=empresa_id, purchase_order_id=purchase_order_id, order_id=order_id)
        return {"items": rows}
    finally:
        svc.close()


@router.get("/orders")
async def list_purchase_orders(
    empresa_id: Optional[int] = Query(None),
    status: Optional[str] = Query(None),
    data_inicio: Optional[str] = Query(None, description="YYYY-MM-DD"),
    data_fim: Optional[str] = Query(None, description="YYYY-MM-DD"),
    invoice_state: Optional[str] = Query(
        None,
        description="Estado fatura/conciliação: fatura_em_falta, por_enviar_cc, na_conta_corrente, conciliado",
    ),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    with_invoice_status: bool = Query(False, description="Incluir estado fatura e has_ledger_entry"),
):
    """Lista ordens de compra (purchase_orders). Se with_invoice_status=true, inclui invoice_ref, invoice_amount, invoice_status, has_ledger_entry e is_reconciled."""
    svc = PurchaseOrderService()
    try:
        if with_invoice_status:
            items, total = svc.list_purchase_orders_with_invoice_status(
                empresa_id=empresa_id,
                status=status,
                invoice_state=invoice_state,
                limit=limit,
                offset=offset,
            )
        else:
            items, total = svc.list_purchase_orders(
                empresa_id=empresa_id, status=status,
                data_inicio=data_inicio, data_fim=data_fim,
                limit=limit, offset=offset,
            )
        return {"items": items, "total": total, "limit": limit, "offset": offset}
    finally:
        svc.close()


# Rotas com segmento extra primeiro (evitar 405 por match em GET /orders/{id})
@router.get("/orders/{purchase_order_id}/checkout-detail")
async def get_checkout_detail(
    purchase_order_id: int,
    portes: Optional[float] = Query(None),
    taxas_pagamento: Optional[float] = Query(None),
):
    """v3.5: Detalhe para Checkout Inteligente (custos, IVA por regime, margem, moradas)."""
    svc = PurchaseCheckoutService()
    try:
        detail = svc.get_checkout_detail(purchase_order_id, portes_override=portes, taxas_pagamento_override=taxas_pagamento)
        if not detail:
            raise HTTPException(status_code=404, detail="Ordem de compra não encontrada")
        return detail
    finally:
        svc.close()


class PoTotalsBody(BaseModel):
    portes_totais: float = 0
    taxas_pagamento: float = 0
    total_final: Optional[float] = None
    valor_base_artigos: Optional[float] = None
    iva_total: Optional[float] = None


@router.patch("/orders/{purchase_order_id}/totals")
async def update_po_totals(purchase_order_id: int, body: PoTotalsBody):
    """v3.5: Atualiza totais da PO (portes, taxas, total; opcionalmente valor base e IVA quando fornecedor alterado)."""
    svc = PurchaseCheckoutService()
    try:
        result = svc.update_po_totals(
            purchase_order_id,
            body.portes_totais,
            body.taxas_pagamento,
            total_final_override=body.total_final,
            valor_base_override=body.valor_base_artigos,
            iva_total_override=body.iva_total,
        )
        if not result.get("success"):
            raise HTTPException(status_code=400, detail=result.get("error", "Erro ao atualizar"))
        return result
    finally:
        svc.close()


class SetSupplierBody(BaseModel):
    supplier_id: int


@router.patch("/orders/{purchase_order_id}/supplier")
async def set_po_supplier(purchase_order_id: int, body: SetSupplierBody):
    """Altera o fornecedor da PO (para o checkout: escolher outro fornecedor que não o proposto)."""
    svc = PurchaseCheckoutService()
    try:
        result = svc.set_po_supplier(purchase_order_id, body.supplier_id)
        if not result.get("success"):
            raise HTTPException(status_code=400, detail=result.get("error", "Erro ao alterar fornecedor"))
        return result
    finally:
        svc.close()


class FinalizePoBody(BaseModel):
    supplier_order_id: Optional[str] = None
    portes_totais: Optional[float] = None
    taxas_pagamento: Optional[float] = None
    total_final: Optional[float] = None
    valor_base_artigos: Optional[float] = None
    iva_total: Optional[float] = None


@router.post("/orders/{purchase_order_id}/finalize")
async def finalize_po(purchase_order_id: int, body: Optional[FinalizePoBody] = None):
    """v3.5: Grava totais (opcional; incl. base e IVA quando fornecedor alterado), regista ID fornecedor e move PO para Ordered."""
    svc = PurchaseCheckoutService()
    try:
        b = body or FinalizePoBody()
        result = svc.finalize_po(
            purchase_order_id,
            supplier_order_id=b.supplier_order_id,
            portes_totais=b.portes_totais,
            taxas_pagamento=b.taxas_pagamento,
            total_final=b.total_final,
            valor_base_artigos=b.valor_base_artigos,
            iva_total=b.iva_total,
        )
        return result
    finally:
        svc.close()


class RegisterInvoiceBody(BaseModel):
    """Registar fatura na PO e opcionalmente enviar para conta corrente."""
    invoice_ref: Optional[str] = None
    invoice_amount: Optional[float] = None
    post_to_ledger: bool = True


@router.patch("/orders/{purchase_order_id}/invoice")
async def register_po_invoice(purchase_order_id: int, body: RegisterInvoiceBody):
    """
    Regista ref e valor da fatura na PO. Se post_to_ledger=true, cria lançamento
    'Fatura' na conta corrente do fornecedor (se ainda não existir).
    """
    po_svc = PurchaseOrderService()
    try:
        updated = po_svc.update_invoice(
            purchase_order_id,
            invoice_ref=body.invoice_ref,
            invoice_amount=body.invoice_amount,
        )
        if not updated:
            raise HTTPException(status_code=404, detail="Ordem de compra não encontrada")
        ledger_created = False
        if body.post_to_ledger:
            acct = AccountingMatchService()
            try:
                existing = acct.conn.execute(
                    "SELECT 1 FROM supplier_ledger WHERE purchase_order_id = ? AND tipo = 'Fatura' LIMIT 1",
                    [purchase_order_id],
                ).fetchone()
                if not existing:
                    valor = body.invoice_amount if body.invoice_amount is not None else updated["total_final"]
                    acct.create_ledger_entry(
                        empresa_id=updated["empresa_id"],
                        supplier_id=updated["supplier_id"],
                        tipo="Fatura",
                        valor_credito=float(valor or 0),
                        documento_ref=body.invoice_ref or f"PO #{purchase_order_id}",
                        purchase_order_id=purchase_order_id,
                        notas="Fatura registada manualmente (match PO)",
                    )
                    ledger_created = True
            finally:
                acct.close()
        return {
            "success": True,
            "purchase_order_id": purchase_order_id,
            "invoice_ref": body.invoice_ref,
            "invoice_amount": body.invoice_amount,
            "ledger_created": ledger_created,
        }
    finally:
        po_svc.close()


class UpdateStatusBody(BaseModel):
    status: str


@router.patch("/orders/{purchase_order_id}")
async def update_purchase_order_status(purchase_order_id: int, body: UpdateStatusBody):
    """Atualiza status da ordem de compra (Draft, Ordered, Paid)."""
    svc = PurchaseOrderService()
    try:
        ok = svc.update_status(purchase_order_id, body.status)
        if not ok:
            raise HTTPException(status_code=400, detail="Status inválido. Use Draft, Ordered ou Paid.")
        return {"success": True, "status": body.status}
    finally:
        svc.close()


@router.get("/orders/{purchase_order_id}")
async def get_purchase_order(purchase_order_id: int):
    """Detalhe de uma ordem de compra com itens."""
    svc = PurchaseOrderService()
    try:
        po = svc.get_purchase_order_detail(purchase_order_id)
        if not po:
            raise HTTPException(status_code=404, detail="Ordem de compra não encontrada")
        return po
    finally:
        svc.close()


@router.get("/pending-sales")
async def get_pending_sales(empresa_id: int = Query(...), limit: int = Query(500, ge=1, le=1000)):
    """Vendas (orders) que ainda não estão em nenhuma ordem de compra (por empresa)."""
    svc = PurchaseOrderService()
    try:
        items = svc.get_pending_sales_orders(empresa_id=empresa_id, limit=limit)
        return {"items": items}
    finally:
        svc.close()


# --- Fase 3: Central de Compras (pending_purchase_items -> POs por empresa/fornecedor) ---


class ConsolidateBody(BaseModel):
    pending_item_ids: List[int]
    portes_totais: float = 0
    taxa_iva_pct: float = 0


@router.get("/pending")
async def get_pending_for_cockpit(
    supplier_id: Optional[int] = Query(None, description="Filtrar por fornecedor"),
    limit: int = Query(2000, ge=1, le=5000),
    service: PurchaseAggregatorService = Depends(get_purchase_aggregator_service),
):
    """Lista consolidada de pending_purchase_items (todas as empresas) para o Cockpit. Agrupar por fornecedor no frontend."""
    try:
        items = service.list_pending_for_cockpit(supplier_id=supplier_id, limit=limit)
        return {"items": items}
    finally:
        service.close()


@router.post("/consolidate")
async def consolidate_purchases(
    body: ConsolidateBody,
    service: PurchaseAggregatorService = Depends(get_purchase_aggregator_service),
):
    """
    Recebe IDs de pending_purchase_items e executa o Split Fiscal.
    Agrupa por (empresa_id, supplier_id) e cria uma PO por grupo. Atualiza status dos itens para 'ordered'.
    """
    try:
        result = service.create_pos_from_pending(
            pending_item_ids=body.pending_item_ids,
            portes_totais=body.portes_totais,
            taxa_iva_pct=body.taxa_iva_pct,
        )
        return result
    finally:
        service.close()


@router.get("/drafts")
async def get_drafts(
    empresa_id: Optional[int] = Query(None),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    service: PurchaseAggregatorService = Depends(get_purchase_aggregator_service),
):
    """Lista de POs em estado Draft para o operador executar no site do fornecedor (Wizard de Checkout)."""
    try:
        items, total = service.list_drafts(empresa_id=empresa_id, limit=limit, offset=offset)
        return {"items": items, "total": total, "limit": limit, "offset": offset}
    finally:
        service.close()


# --- Global Cockpit & Bulk Prepare (múltiplas empresas, baseado em orders) ---

@router.get("/global-pending")
async def get_global_pending(
    supplier_id: Optional[int] = Query(None, description="Filtrar por fornecedor"),
    limit: int = Query(1000, ge=1, le=2000),
    service: BulkPurchaseService = Depends(get_bulk_purchase_service),
):
    """Lista vendas pendentes de todas as empresas, agrupáveis por fornecedor (Global Cockpit)."""
    try:
        items = service.get_global_pending(supplier_id=supplier_id, limit=limit)
        return {"items": items}
    finally:
        service.close()


class PrepareBulkBody(BaseModel):
    order_ids: List[int]
    portes_totais: float = 0
    taxa_iva_pct: float = 0
    tipo_envio: str = "Escritorio"


@router.post("/prepare-bulk")
async def prepare_bulk_purchases(
    body: PrepareBulkBody,
    service: BulkPurchaseService = Depends(get_bulk_purchase_service),
):
    """
    Seleção global -> múltiplas POs (uma por empresa_id).
    Agrupa order_ids por empresa e fornecedor, cria uma PO por grupo com dados fiscais.
    """
    try:
        result = service.prepare_bulk_purchases(
            order_ids=body.order_ids,
            portes_totais=body.portes_totais,
            taxa_iva_pct=body.taxa_iva_pct,
            tipo_envio=body.tipo_envio,
        )
        return result
    finally:
        service.close()


@router.get("/sale-state")
async def get_sale_state(
    empresa_id: Optional[int] = Query(None),
    limit: int = Query(500, ge=1, le=2000),
    service: BulkPurchaseService = Depends(get_bulk_purchase_service),
):
    """BI: estado de cada venda (Vendido | Em Processamento de Compra | Comprado (PO #123))."""
    try:
        items = service.get_sale_state_list(empresa_id=empresa_id, limit=limit)
        return {"items": items}
    finally:
        service.close()


class SetSupplierOrderIdBody(BaseModel):
    supplier_order_id: str


@router.patch("/orders/{purchase_order_id}/supplier-order-id")
async def set_supplier_order_id(
    purchase_order_id: int,
    body: SetSupplierOrderIdBody,
):
    """Regista o ID da encomenda no site do fornecedor (após checkout)."""
    svc = PurchaseOrderService()
    try:
        svc.update_supplier_order_id(purchase_order_id, body.supplier_order_id)
        return {"success": True, "supplier_order_id": body.supplier_order_id}
    finally:
        svc.close()
